src/Microsoft.ML.FastTree/GamTrainer.cs

```diff
@@ -1028,6 +1030,118 @@ void ICanSaveSummary.SaveSummary(TextWriter writer, RoleMappedSchema schema)
             Runtime.Numeric.VectorUtils.SparsifyNormalize(ref contributions, top, bottom, normalize);
         }
 
+        public void SaveAsIni(TextWriter writer, RoleMappedSchema schema, ICalibrator calibrator = null)
+        {
+            Host.CheckValue(writer, nameof(writer));
+            Host.CheckValue(schema, nameof(schema));
+            Host.CheckValueOrNull(calibrator);
+
+            var ensemble = new TreeEnsemble();
+
+            for (int featureIndex = 0; featureIndex < _numFeatures; featureIndex++)
+            {
+                var effects = _binEffects[featureIndex];
+                var binThresholds = _binUpperBounds[featureIndex];
+
+                Host.Assert(effects.Length == binThresholds.Length);
+                var numLeaves = effects.Length;
+                var numInternalNodes = numLeaves - 1;
+
+                var splitFeatures = Enumerable.Repeat(featureIndex, numInternalNodes).ToArray();
+                var (treeThresholds, lteChild, gtChild) = CreateBalancedTree(numInternalNodes, binThresholds);
+                var tree = CreateRegressionTree(numLeaves, splitFeatures, treeThresholds, lteChild, gtChild, effects);
+                ensemble.AddTree(tree);
+            }
+
+            // Tried adding the intercept as the bias term in the final ini aggregator,
+            // but that didn't seem to have any effects during testing.
+            // Adding the intercept as a dummy tree with the output values being the model intercept,
+            // works for reaching parity.
+            var interceptTree = CreateRegressionTree(
+                numLeaves: 2,
+                splitFeatures: new[] { 0 },
+                rawThresholds: new[] { 0f },
+                lteChild: new[] { ~0 },
+                gtChild: new[] { ~1 },
+                leafValues: new[] { Intercept, Intercept });
+            ensemble.AddTree(interceptTree);
+
+            var ini = FastTreeIniFormatUtils.TreeEnsembleToIni(
+                Host, ensemble, schema, calibrator, string.Empty, false, false);
+
+            // Remove the SplitGain values which are all 0.
+            // It's eaiser to remove them here, than to modify the FastTree code.
+            var goodLines = ini.Split(new[] { '\n' }).Where(line => !line.StartsWith("SplitGain="));
+            ini = string.Join("\n", goodLines);
+            writer.WriteLine(ini);
+        }
+
+        // GAM bins should be converted to balanced trees / binary search trees
+        // so that scoring takes O(log(n)) instead of O(n). The following utility
+        // creates a balanced tree.
+        private (float[], int[], int[]) CreateBalancedTree(int numInternalNodes, double[] binThresholds)
```

Comment:
Does returning a list limit what C# or .NET versions we can target? If so, we can switch this to an `out` pattern.

Reply:
Tuples are heavily used as return type in Static API, so using it here doesn't create more version restrictions than what we already have in the rest of the code. I think this way it fine.

---
In reply to: [243140927](https://github.com/dotnet/machinelearning/pull/1929#discussion_r243140927) [](ancestors = 243140927)